// @civic/solana-gateway-ts/dist/lib/solanaBorsh.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Enum = exports.Assignable = exports.SCHEMA = void 0;
const borsh_1 = require("borsh");
exports.SCHEMA = new Map();
// Class wrapping a plain object
class Assignable {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    constructor(properties) {
        Object.keys(properties).forEach((key) => {
            // this is probably possible in Typescript,
            // but requires (keyof this) which is not possible in the constructor
            // WARNING: This functionality actually changes depending on which ts compiler you use
            // it works with tsc, not with bun. To be safe we should change it.
            // @ts-expect-error - therefore we need to disable this error.
            this[key] = properties[key];
        });
    }
    encode() {
        return Buffer.from((0, borsh_1.serialize)(exports.SCHEMA, this));
    }
    static decode(data) {
        // use deserializeUnchecked here as opposed to deserialize,
        // as the latter throws an error if the data has trailing bytes.
        return (0, borsh_1.deserializeUnchecked)(exports.SCHEMA, this, data);
    }
}
exports.Assignable = Assignable;
// Class representing a Rust-compatible enum, since enums are only strings or
// numbers in pure JS
class Enum extends Assignable {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    constructor(properties) {
        super(properties);
        if (Object.keys(properties).length !== 1) {
            throw new Error("Enum can only take single value");
        }
        this.enum = "";
        Object.keys(properties).forEach((key) => {
            this.enum = key;
        });
    }
}
exports.Enum = Enum;
