// @civic/gateway-client-react/dist/cjs/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ButtonMode = exports.getTokenDescription = exports.IdentityButton = exports.GatewayStatus = exports.GatewayProvider = exports.useGateway = void 0;
var GatewayClientContext_1 = require("./GatewayClientContext");
Object.defineProperty(exports, "useGateway", { enumerable: true, get: function () { return GatewayClientContext_1.useGateway; } });
Object.defineProperty(exports, "GatewayProvider", { enumerable: true, get: function () { return GatewayClientContext_1.GatewayProvider; } });
var types_1 = require("./types");
Object.defineProperty(exports, "GatewayStatus", { enumerable: true, get: function () { return types_1.GatewayStatus; } });
var button_1 = require("./button");
Object.defineProperty(exports, "IdentityButton", { enumerable: true, get: function () { return button_1.IdentityButton; } });
Object.defineProperty(exports, "getTokenDescription", { enumerable: true, get: function () { return button_1.getTokenDescription; } });
Object.defineProperty(exports, "ButtonMode", { enumerable: true, get: function () { return button_1.ButtonMode; } });
