// @civic/solana-gateway-chain-client/dist/esm/config.js
export const makeConfig = (clusterUrl, cluster) => {
    return {
        cluster,
        commitment: 'confirmed',
        // this map instructs the POWO library to use clusterUrl
        // to connect to the solana network. This avoids rate limiting issues with using the default
        // public urls
        supportedClusterUrls: {
            [cluster]: clusterUrl,
        },
        recentBlockCheck: false,
        broadcastCheck: false,
    };
};
/**
 * Return the payer based on the gatekeeperSendsTransaction flag and any specified payer.
 * If gatekeeperSendsTransaction is defined, then ownerSings is the inverse of its value
 * If gatekeeperSendsTransaction is not defined, then ownerSings is the inverse of broadcastTransaction
 * If neither are defined, ownerSigns is enabled
 */
export const getPayer = ({ gatekeeperSendsTransaction, payerParameter, owner }) => {
    if (gatekeeperSendsTransaction)
        return undefined; // Gatekeeper is the sender, no payer is passed from the client
    // Payer is passed from the client or payer and gatekeeperSendsTransaction are both unspecified
    // If payer is not passed, use the wallet address if gatekeeperSendsTransaction is not set or false
    return payerParameter || owner;
};
