// @civic/solana-gateway-react/dist/cjs/index.js
"use strict";
/* eslint-disable no-console */
/* eslint-disable global-require */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SolanaChainImplementation = exports.GatewayProvider = exports.useGateway = exports.Badge = exports.ErrorCode = exports.ButtonMode = exports.getTokenDescription = exports.IdentityButton = exports.GatewayStatus = void 0;
const version_1 = require("./version");
var gateway_client_react_1 = require("@civic/gateway-client-react");
Object.defineProperty(exports, "GatewayStatus", { enumerable: true, get: function () { return gateway_client_react_1.GatewayStatus; } });
Object.defineProperty(exports, "IdentityButton", { enumerable: true, get: function () { return gateway_client_react_1.IdentityButton; } });
Object.defineProperty(exports, "getTokenDescription", { enumerable: true, get: function () { return gateway_client_react_1.getTokenDescription; } });
Object.defineProperty(exports, "ButtonMode", { enumerable: true, get: function () { return gateway_client_react_1.ButtonMode; } });
var gateway_client_core_1 = require("@civic/gateway-client-core");
Object.defineProperty(exports, "ErrorCode", { enumerable: true, get: function () { return gateway_client_core_1.ErrorCode; } });
var badge_1 = require("./badge");
Object.defineProperty(exports, "Badge", { enumerable: true, get: function () { return __importDefault(badge_1).default; } });
var useSolanaGateway_1 = require("./useSolanaGateway");
Object.defineProperty(exports, "useGateway", { enumerable: true, get: function () { return useSolanaGateway_1.useSolanaGateway; } });
var SolanaGatewayProvider_1 = require("./SolanaGatewayProvider");
Object.defineProperty(exports, "GatewayProvider", { enumerable: true, get: function () { return SolanaGatewayProvider_1.SolanaGatewayProvider; } });
var solana_gateway_chain_client_1 = require("@civic/solana-gateway-chain-client");
Object.defineProperty(exports, "SolanaChainImplementation", { enumerable: true, get: function () { return solana_gateway_chain_client_1.SolanaChainImplementation; } });
console.log(version_1.VERSION);
