// @civic/solana-gateway-react/dist/cjs/SolanaGatewayProvider.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.useSolanaGateway = exports.SolanaGatewayProvider = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const gateway_client_react_1 = require("@civic/gateway-client-react");
const gateway_client_core_1 = require("@civic/gateway-client-core");
const solana_gateway_chain_client_1 = require("@civic/solana-gateway-chain-client");
const version_1 = require("./version");
const logInfo = (0, gateway_client_core_1.prefixLogger)('SolanaGatewayProvider').info;
const SolanaGatewayProvider = ({ children = null, wallet, connection, cluster = 'mainnet-beta', gatekeeperNetwork, wrapper, logo, stage = 'prod', redirectUrl, gatekeeperSendsTransaction, payer, handleTransaction, expiryMarginSeconds, options, partnerAppId, forceRequireRefresh, hideWalletPrompts, }) => {
    var _a;
    // default to true in case the passed wallet doesn't support the 'connected' property
    if (payer && gatekeeperSendsTransaction)
        throw Error('Cannot set payer when gatekeeperSendsTransaction is true');
    // these values are dependant on one-another so they need to all change together if an input changes
    const gatewayProviderProps = (0, react_1.useMemo)(() => {
        var _a;
        return ({
            walletAddress: (_a = wallet === null || wallet === void 0 ? void 0 : wallet.publicKey) === null || _a === void 0 ? void 0 : _a.toBase58(),
            chainImplementation: (wallet === null || wallet === void 0 ? void 0 : wallet.publicKey) && (wallet === null || wallet === void 0 ? void 0 : wallet.signTransaction) && stage && (gatekeeperNetwork === null || gatekeeperNetwork === void 0 ? void 0 : gatekeeperNetwork.toBase58())
                ? new solana_gateway_chain_client_1.SolanaChainImplementation({
                    connection,
                    cluster,
                    publicKey: wallet.publicKey,
                    signTransaction: wallet.signTransaction,
                    signMessage: wallet.signMessage,
                    gatekeeperNetworkAddress: gatekeeperNetwork,
                    stage,
                    handleTransaction,
                })
                : undefined,
            stage,
            gatekeeperNetwork: gatekeeperNetwork === null || gatekeeperNetwork === void 0 ? void 0 : gatekeeperNetwork.toBase58(),
            options: Object.assign(Object.assign({}, options), { 
                // solana blockhashes are short so a default of only 1 client-side retry is reasonable
                clientSendsMaxRetries: (options === null || options === void 0 ? void 0 : options.clientSendsMaxRetries) || 1 }),
            gatekeeperSendsTransaction,
            expiryMarginSeconds,
            partnerAppId,
            disableInitialGatekeeperLookup: [
                'tigoYhp9SpCDoCQmXGj2im5xa3mnjR1zuXrpCJ5ZRmi',
                'ignREusXmGrscGNUesoU9mxfds9AiYTezUKex2PsZV6',
            ].includes((gatekeeperNetwork === null || gatekeeperNetwork === void 0 ? void 0 : gatekeeperNetwork.toBase58()) || ''),
            payer: payer === null || payer === void 0 ? void 0 : payer.toBase58(),
            hideWalletPrompts,
        });
    }, [
        (_a = wallet === null || wallet === void 0 ? void 0 : wallet.publicKey) === null || _a === void 0 ? void 0 : _a.toBase58(),
        connection === null || connection === void 0 ? void 0 : connection.rpcEndpoint,
        cluster,
        stage,
        gatekeeperNetwork === null || gatekeeperNetwork === void 0 ? void 0 : gatekeeperNetwork.toBase58(),
        handleTransaction,
        JSON.stringify(options),
        gatekeeperSendsTransaction,
        expiryMarginSeconds,
        partnerAppId,
        payer,
        hideWalletPrompts,
    ]);
    (0, react_1.useEffect)(() => {
        logInfo('Client Options', Object.assign(Object.assign({}, gatewayProviderProps), { forceRequireRefresh }));
    }, [gatewayProviderProps, forceRequireRefresh]);
    return ((0, jsx_runtime_1.jsx)(gateway_client_react_1.GatewayProvider
    // memoised values that need to be synchronised
    , Object.assign({ 
        // memoised values that need to be synchronised
        walletAddress: gatewayProviderProps.walletAddress, chainImplementation: gatewayProviderProps.chainImplementation, stage: gatewayProviderProps.stage, gatekeeperNetwork: gatewayProviderProps.gatekeeperNetwork, options: gatewayProviderProps.options, gatekeeperSendsTransaction: !!gatewayProviderProps.gatekeeperSendsTransaction, expiryMarginSeconds: gatewayProviderProps.expiryMarginSeconds, partnerAppId: gatewayProviderProps.partnerAppId, disableInitialGatekeeperLookup: gatewayProviderProps.disableInitialGatekeeperLookup, payer: gatewayProviderProps.payer, 
        // non-memoised values that don't need to be synchronised
        redirectUrl: redirectUrl, logo: logo, 
        // Sent to the Gatekeeper in the x-civic-client version header.
        // If this is not passed in, the version of the generic gateway-client-react package will be used.
        reactComponentVersion: version_1.VERSION, wrapper: wrapper, forceRequireRefresh: forceRequireRefresh, expectTokenTimeoutSeconds: 30 }, { children: children })));
};
exports.SolanaGatewayProvider = SolanaGatewayProvider;
var useSolanaGateway_1 = require("./useSolanaGateway");
Object.defineProperty(exports, "useSolanaGateway", { enumerable: true, get: function () { return useSolanaGateway_1.useSolanaGateway; } });
__exportStar(require("./types"), exports);
