// @civic/solana-gateway-react/dist/cjs/useSolanaGateway.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useSolanaGateway = void 0;
const web3_js_1 = require("@solana/web3.js");
const gateway_client_react_1 = require("@civic/gateway-client-react");
const util_1 = require("./util");
const useSolanaGateway = () => {
    const { gatewayToken, gatewayTokenTransaction } = (0, gateway_client_react_1.useGateway)();
    const solanaGatewayToken = gatewayToken
        ? {
            issuingGatekeeper: new web3_js_1.PublicKey(gatewayToken.issuingGatekeeper),
            gatekeeperNetworkAddress: new web3_js_1.PublicKey(gatewayToken.gatekeeperNetworkAddress),
            owner: new web3_js_1.PublicKey(gatewayToken.owner),
            state: gatewayToken.state,
            publicKey: new web3_js_1.PublicKey(gatewayToken.identifier),
            expiryTime: gatewayToken.expiryTime,
        }
        : undefined;
    const solanaGatewayTokenTransaction = gatewayTokenTransaction
        ? (0, util_1.createSolanaTransactionFromBase64)(gatewayTokenTransaction)
        : undefined;
    return Object.assign(Object.assign({}, (0, gateway_client_react_1.useGateway)()), { gatewayToken: solanaGatewayToken, gatewayTokenTransaction: solanaGatewayTokenTransaction });
};
exports.useSolanaGateway = useSolanaGateway;
